aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -1734,6 +1735,9 @@ Tensor masked_select_cpu(const Tensor & self, const Tensor & mask) {
 }
 
 Tensor masked_select_backward(const Tensor& grad, const Tensor& input, const Tensor& mask) {
+  if (areAnyTensorSubclassLike({grad, input, mask})) {
+    return at::zeros_like(input, input.options()).masked_scatter(mask, grad);
+  }
```

Comment:
Wouldn't this fail in the case mentioned below?
I think the correct fix would be to simply use `masked_scatter` rather than `masked_scatter_` below.
Otherwise, use a ternary operator to choose between the in-place and the out-of-place in the last line.

Reply:
For the below mentioned case, we will always hit the `if`. Ternary seemed bit clunky. Let me know if ternary seems better :)

Reply:
is the issue here about broadcasting @Lezcano? Seems like both paths seem to handle that?

Reply:
how does this new path handle it?

Reply:
(note that `mask` operations do not broadcast)

Reply:
Not sure if intended but `masked_scatter` takes care of broadcasting the LHS.

https://github.com/pytorch/pytorch/blob/417676720e3bf9e77b70eeaff5595f4c80a990e6/aten/src/ATen/native/TensorAdvancedIndexing.cpp#L1571-L1575

Reply:
Há! How about that!

Then, if we care about efficiency, I'd implement it as
```c++
auto ret = at::zeros_like(input);
if (areAnyTensorSubclassLike({grad, input, mask})) {
  return ret.masked_scatter(mask, grad);
} else {
  ret.masked_scatter_(mask, grad);
  return ret;
}
```
if we want to go fancy (we do not return `ret.masked_scatter_` directly because it returns a `Tensor&` and it'd incur a (shallow) copy of `Tensor`. 

On the other end, if we put readability first, this could be written as an inline formula in `derivatives.yaml` as:
```c++
  zeros_like(self).masked_scatter(mask, grad)
```

Reply:
Thanks! Will stick go for the first one with `if` :)